code/go/0chain.net/blobbercore/allocation/allocationchange.go

```diff
@@ -33,33 +34,56 @@ type AllocationChangeProcessor interface {
 }
 
 type AllocationChangeCollector struct {
-	ConnectionID      string                      `gorm:"column:connection_id;primary_key"`
-	AllocationID      string                      `gorm:"column:allocation_id"`
-	ClientID          string                      `gorm:"column:client_id"`
-	Size              int64                       `gorm:"column:size"`
-	Changes           []*AllocationChange         `gorm:"ForeignKey:connection_id;AssociationForeignKey:connection_id"`
+	ConnectionID      string                      `gorm:"column:connection_id;primaryKey"`
+	AllocationID      string                      `gorm:"column:allocation_id;size:64;not null"`
+	ClientID          string                      `gorm:"column:client_id;size:64;not null"`
+	Size              int64                       `gorm:"column:size;not null;default:0"`
+	Changes           []*AllocationChange         `gorm:"foreignKey:ConnID"`
 	AllocationChanges []AllocationChangeProcessor `gorm:"-"`
-	Status            int                         `gorm:"column:status"`
+	Status            int                         `gorm:"column:status;not null;default:0"`
 	datastore.ModelWithTS
 }
 
 func (AllocationChangeCollector) TableName() string {
 	return "allocation_connections"
 }
 
+func (ac *AllocationChangeCollector) BeforeCreate(tx *gorm.DB) error {
+	ac.CreatedAt = time.Now()
+	ac.UpdatedAt = ac.CreatedAt
+	return nil
+}
+
+func (ac *AllocationChangeCollector) BeforeSave(tx *gorm.DB) error {
+	ac.UpdatedAt = time.Now()
+	return nil
+}
+
 type AllocationChange struct {
-	ChangeID     int64  `gorm:"column:id;primary_key"`
-	Size         int64  `gorm:"column:size"`
-	Operation    string `gorm:"column:operation"`
-	ConnectionID string `gorm:"column:connection_id"`
-	Input        string `gorm:"column:input"`
+	ChangeID   int64                     `gorm:"column:id;primaryKey"`
+	Size       int64                     `gorm:"column:size;not null;default:0"`
+	Operation  string                    `gorm:"column:operation;size:20;not null"`
+	ConnID     string                    `gorm:"column:connection_id;size:64;not null"`
+	Connection AllocationChangeCollector `gorm:"foreignKey:ConnID"` // References allocation_connections(connection_id)
```

Comment:
Seems `AlloocationChangeCollector` does not have `ConnID` field, so perhaps `ConnectionID`

Reply:
@peterlimg We cannot have `ConnectionID` as a field name when using gorm's automigration. It just doesn't work. This is issue with gorm but there is no other workaround.

But your concern with `ConnID` field; Gorm can create reference to the `allocation_connections` table. It actually is the gorm's tag that matters. And for gorm tag we have column name as `connection_id`.

Reply:
Isn't the `foreignKey:ConnID` reference to the primary key defined in `AllocationChangeCollector` struct, which has `ConnectionID string` field with `gorm:"column:connection_id;primaryKey"` tag? I don't see the `ConnID` defined here, so I thought it mistakenly set it as `ConnID`. Anyway, do think the `foreignKey:ConnID` could reference it?

Reply:
If we use `ConnectionID`, automigration will panic.

`ConnID` is not needed to define in `AllocationChangeCollector` struct, but if you see `gorm's` tag for `ConnID`, its `connection_id`. 
It will only then create proper reference between tables.

Reply:
Yes, I got what happens here, perhaps if we rename it to `AllocationChangeCollectorRefer` or `ConnectionRefer` would make it easier to understand. Check the tables.go I changed here: https://gist.github.com/peterlimg/2dbbb6916b1a46e7219c7cdd2e6880c4